HeterogeneousCore/AlpakaCore/interface/alpaka/Event.h

```diff
@@ -0,0 +1,152 @@
+#ifndef HeterogeneousCore_AlpakaCore_interface_alpaka_Event_h
+#define HeterogeneousCore_AlpakaCore_interface_alpaka_Event_h
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/EDPutToken.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/DeviceProductType.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/EDGetToken.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/EDMetadata.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/EDPutToken.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE::device {
+  /**
+   * The device::Event mimics edm::Event, and provides access to
+   * EDProducts in the host memory space, and in the device memory
+   * space defined by the backend (i.e. ALPAKA_ACCELERATOR_NAMESPACE).
+   * The device::Event also gives access to the Queue object the
+   * EDModule code should use to queue all the device operations.
+   *
+   * Access to device memory space products is synchronized properly.
+   * For backends with synchronous Queue this is trivial. For
+   * asynchronous Queue, either the Queue of the EDModule is taken
+   * from the first data product, or a wait is inserted into the
+   * EDModule's Queue to wait for the product's asynchronous
+   * production to finish.
+   *
+   * Note that not full interface of edm::Event is replicated here. If
+   * something important is missing, that can be added.
+   */
+  class Event {
+  public:
+    // To be called in produce()
+    explicit Event(edm::Event& ev, std::shared_ptr<EDMetadata> metadata)
+        : constEvent_(ev), event_(&ev), metadata_(std::move(metadata)) {}
+
+    // To be called in acquire()
+    explicit Event(edm::Event const& ev, std::shared_ptr<EDMetadata> metadata)
+        : constEvent_(ev), metadata_(std::move(metadata)) {}
+
+    Event(Event const&) = delete;
+    Event& operator=(Event const&) = delete;
+    Event(Event&&) = delete;
+    Event& operator=(Event&&) = delete;
+
+    auto streamID() const { return constEvent_.streamID(); }
+    auto id() const { return constEvent_.id(); }
+
+    // Alpaka operations do not accept a temporary as an argument
+    // TODO: Returning non-const reference here is BAD
+    Queue& queue() const {
+      queueUsed_ = true;
+      return metadata_->queue();
+    }
+
+    // Note: this function is planned to be removed after a more
+    // proper treatment of multiple devices per backend is implemented
+    // in EventSetup
+    Device device() const { return alpaka::getDev(metadata_->queue()); }
```

Comment:
Mhm, I think it may actually be useful to provide the underlying device to the `EDProducer`s.
For example, if I remember correctly CUDA devices should be able provide information about the corresponding hardware device, like the SM or the number of threads, that an `EDProducer` could use to tune the kernel parameters.

Reply:
Ok. I'll then organize the code a bit differently (essentially make the `Device` available already from `EDMetadata`).

In principle user could always ask the device from the `Queue` with `alpaka::getDev()`. But if a user asks the `Queue` before reading the first device-side data product, there is no attempt to re-use the `Queue` from that data product. So it is probably worth to provide an explicit function to access the `Device`.